core/txindexer.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>
+
+package core
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ava-labs/subnet-evm/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// TxIndexProgress is the struct describing the progress for transaction indexing.
+type TxIndexProgress struct {
+	Indexed   uint64 // number of blocks whose transactions are indexed
+	Remaining uint64 // number of blocks whose transactions are not indexed yet
+}
+
+// Done returns an indicator if the transaction indexing is finished.
+func (progress TxIndexProgress) Done() bool {
+	return progress.Remaining == 0
+}
+
+// txIndexer is the module responsible for maintaining transaction indexes
+// according to the configured indexing range by users.
+type txIndexer struct {
+	// limit is the maximum number of blocks from head whose tx indexes
+	// are reserved:
+	//  * 0: means the entire chain should be indexed
+	//  * N: means the latest N blocks [HEAD-N+1, HEAD] should be indexed
+	//       and all others shouldn't.
+	limit    uint64
+	db       ethdb.Database
+	progress chan chan TxIndexProgress
+	term     chan chan struct{}
+	closed   chan struct{}
+
+	chain *BlockChain
+}
+
+// newTxIndexer initializes the transaction indexer.
+func newTxIndexer(limit uint64, chain *BlockChain) *txIndexer {
+	indexer := &txIndexer{
+		limit:    limit,
+		db:       chain.db,
+		progress: make(chan chan TxIndexProgress),
+		term:     make(chan chan struct{}),
+		closed:   make(chan struct{}),
+		chain:    chain,
+	}
+	chain.wg.Add(1)
+	var (
+		headCh = make(chan ChainEvent, 1) // XXX: Buffered to avoid locking up the event feed
```

Comment:
I think this is same as the previous one, so should be fine

Reply:
I left it as is for now, but does the channel need to be buffered (I think it's not buffered upstream)?

Reply:
it was buffered until the code refactored into its own file. txpool does not use buffered channel, but in other places a buffered channel is used for such subscriptions. Any possibility they forgot to add the buffered channel during refactor? Because it seems that it could actually lock it up until indexer finishes the `run`.

Reply:
Yes you are right about that it was changed during the refactor, seems like we should figure it out and make a PR to upstream if needed.

Reply:
https://github.com/ethereum/go-ethereum/pull/30350

Reply:
we got a response, seems we don't need a buffered ch: https://github.com/ethereum/go-ethereum/pull/30350

I totally forgot we run the indexer with another gouroutine, so should be good to use a unbuffered ch.